### ISSUE TITLE
Promises don't work

### ISSUE THREAD

**Issue by clayrisser** [jamrizzi/gtk-node#4]
Promises appear to break ffi. My temporary solution is to use old fashioned callbacks. If anyone had any insight on this, please let me know. I've filled an issue to ffi at the link below.

https://github.com/node-ffi/node-ffi/issues/412

**Comment by Place1** [comment]
So I think this could be an issue caused by a blocked Javascript event loop.
The `app.init()` call hits `g_application_run()` [see source](https://github.com/Place1/node-gtk3/blob/experimenting/src/app/index.c#L34) and this C function blocks until the UI is closed. Because a Javascript function is calling into this, that javascript function `app.init()` will be blocked as well. Because Node is single threaded this stops all Node code, such as pending promises or setTimeout callbacks, from running!

Take this example:

```
import { App, Button } from '../../lib';

const app = new App({
  title: 'Node Gtk'
});
app.init((window) => {
  const button = new Button({
    name: 'Button 1'
  });
  button.attach(window);
  console.log('creating a set timeout');
  setTimeout(() => console.log('this will run after the application window closes'), 1000);
  app.render();
});

console.log('I will run after the application window is closed!');
```
Output:
```
creating a set timeout
// nothing will happen...
// now after the UI is closed by the user...
I will run after the application window is closed!
this will run 1 second after the application window closes
```

**Comment by clayrisser** [comment]
Wow, super inciteful. This can definitely be fixed.

**Comment by clayrisser** [comment]
I fixed it with c75dab5cbcd25d5d94bbce8537b18ecc49d68f19

I'm now using promises instead of callbacks

**Comment by Place1** [comment]
I'm not sure if this commit fixes the issue. I'm getting a seg-fault running the demo now but besides that, there's still a blocking piece of C code.

[source](https://github.com/jamrizzi/node-gtk3/blob/master/src/app/index.c#L37)
```
int init(GtkApplication *app) {
  g_signal_connect(app, "activate", G_CALLBACK(activate), NULL);
  int status = g_application_run(G_APPLICATION(app), 0, NULL);
  g_object_unref(app);
  return status;
}
```

the `g_application_run(G_APPLICATION(app), 0, NULL);` method blocks indefinetly until the UI is closed.

EDIT:
in regards to the seg fault, I put a 'console.log()'  at the start of the demo and that appears to result in a different error:
```
/home/james/react-gtk3-project/node-gtk3/node_modules/ffi/lib/callback.js:18
        throw new Error(err)
        ^

Error: ffi fatal: callback has been garbage collected!
    at /home/james/react-gtk3-project/node-gtk3/node_modules/ffi/lib/callback.js:18:15
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```
So i guess there's some sort of race condition happening.

**Comment by Place1** [comment]
i'm working on a fix

**Comment by Place1** [comment]
https://github.com/Place1/node-gtk3/commit/22ac86bd40776855b52ce58caa956a9a9d03d875
https://github.com/Place1/node-gtk3/commit/134758b07cb4e590867bd0a97b7c641d4908fb4f

This is were i'm up to. Promises, callbacks, setTimouts etc. all work now, but the program seg faults after a few seconds unfortuantly. I've added some comments on why I think this happens

```
init() {
    loop.startLoop();
    return new Promise((resolve, reject) => {
      // assign the ffi callback to 'this' so we retain a reference
      // after this.init() finishes. If we don't do this Node
      // will garbage collect it and we will get a segmentation
      // fault when C calls into it.
      this.onActivate = ffi.Callback('void', [type.GtkWidgetPtr], (windowPtr) => {
        this.window = new Window({ pointer: windowPtr });
        resolve(this.window);
      });
      app.register_on_activate(this.onActivate);

      // If we call app.init() then it blocks the main loop
      // forever and out program halts.
      // If we call app.init.async() then FFI will call the
      // foreign function in another thread (https://github.com/node-ffi/node-ffi/wiki/Node-FFI-Tutorial#async-library-calls)
      // My guess is that it will be scheduled on one of the
      // threads that LibUV controls for IO tasks.
      // This means we avoid blocking the event loop but this
      // results in a segmentation fault after some time and
      // I can't figure out why. My only guess is that it's a
      // multi-thread issue with GTK???
      app.init.async(this.pointer, () => null);
    });
  }
```

**Comment by clayrisser** [comment]
The fix for the blocking event loop was not a change in the C code. It was a change in how I called the init function.

```js
// blocking app.js
      app.init(this.pointer);
```

```js
// unblocking app.js
      app.init.async(this.pointer, (err, status) => {
        if (status !== 0) {
          console.error(new Error('Program initialized with error'));
        }
        process.exit(status);
      });
```

**Comment by clayrisser** [comment]
Let's move the seg-fault to a new issue.

**Comment by clayrisser** [comment]
I filed a new issue for this at #6
